src/pure.tsx

```diff
@@ -0,0 +1,130 @@
+import React, { ReactElement, ReactNode, Suspense } from 'react'
+import { act, create, ReactTestRenderer } from 'react-test-renderer'
+import asyncUtils from './asyncUtils'
+import { cleanup, addCleanup, removeCleanup } from './cleanup'
+
+type TestHookProps<TProps, TResult> = {
+  callback: (props: TProps) => TResult
+  hookProps: TProps | undefined
+  onError: (error: Error) => void
+  children: (value: TResult) => void
+}
+function TestHook<TProps, TResult>({
+  callback,
+  hookProps,
+  onError,
+  children
+}: TestHookProps<TProps, TResult>) {
+  try {
+    if (hookProps) {
+      children(callback(hookProps))
+    } else {
+      // coerce into undefined, so it maintains the previous behaviour
+      children(callback((undefined as unknown) as TProps))
+    }
+    // eslint-disable-next-line @typescript-eslint/no-implicit-any-catch
+  } catch (err) {
+    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
+    if (err.then) {
+      throw err
+    } else {
+      onError(err)
+    }
+  }
+  return null
+}
```

Comment:
The catch block was bothering me so I tried something... I _think_ it's better (it's definitely more explicit), but would like some feedback on it:

```suggestion
 
function isPromise<T>(value: unknown) : value is PromiseLike<T> {
  return typeof (value as PromiseLike<T>).then === 'function'
}

function TestHook<TProps, TResult>({
  callback,
  hookProps,
  onError,
  children
}: TestHookProps<TProps, TResult>) {
  try {
    if (hookProps) {
      children(callback(hookProps))
    } else {
      // coerce into undefined, so it maintains the previous behaviour
      children(callback((undefined as unknown) as TProps))
    }
  } catch (err: unknown) {
    if (isPromise(err)) {
      throw err as unknown
    } else {
      onError(err as Error)
    }
  }
  return null
}
```

(I'm not sure if you can do suggestions for multiple seperate blocks, so just focus on the `catch` block and `isPromise` function).

Reply:
Why is the promise error being cast back to unknown, after you went to the effort of Type Guarding it? You may as well just not Type Guard it and just keep err as the unknown type.

Reply:
Yeah, it was _conceptually_ more correct than being more useful.  Perhaps leaving the return type as just `boolean` on `isPromise` is a bit easier.